guava/src/com/google/common/hash/BloomFilter.java

```diff
@@ -380,4 +414,72 @@ Object readResolve() {
     }
     private static final long serialVersionUID = 1;
   }
+
+  /**
+   * Writes this {@code BloomFilter} to an output stream, with a custom format (not Java
+   * serialization). This has been measured to save at least 400 bytes compared to regular
+   * serialization.
+   *
+   * <p>Use {@linkplain #readFrom(InputStream, Funnel)} to reconstruct the written BloomFilter.
+   */
+  public void writeTo(OutputStream out) throws IOException {
+    /*
+     * Serial form:
+     * 1 signed byte for the strategy
+     * 1 unsigned byte for the number of hash functions
+     * 1 big endian int, the number of longs in our bitset
+     * N big endian longs of our bitset
+     */
+    DataOutputStream dout = new DataOutputStream(out);
+    dout.writeByte(SignedBytes.checkedCast(strategy.ordinal()));
+    dout.writeByte(UnsignedBytes.checkedCast(numHashFunctions)); // note: checked at the c'tor
+    dout.writeInt(bits.data.length);
+    for (long value : bits.data) {
+      dout.writeLong(value);
+    }
+  }
+
+  /**
+   * Reads a byte stream, which was written by {@linkplain #writeTo(OutputStream)}, into
+   * a {@code BloomFilter<T>}.
+   *
+   * The {@code Funnel} to be used is not encoded in the stream, so it must be provided here.
+   * <b>Warning:</b> the funnel provided <b>must</b> behave identically to the one used to
+   * populate the original Bloom filter!
+   *
+   * @throws IOException if the InputStream throws an {@code IOException}, or if its data does
+   *     not appear to be a BloomFilter serialized using the
+   *     {@linkplain #writeTo(OutputStream)} method.
+   */
+  public static <T> BloomFilter<T> readFrom(InputStream in, Funnel<T> funnel) throws IOException {
+    checkNotNull(in, "InputStream");
+    checkNotNull(funnel, "Funnel");
+    int strategyOrdinal = -1;
+    int numHashFunctions = -1;
+    int dataLength = -1;
+    try {
+      DataInputStream din = new DataInputStream(in);
+      // currently this assumes there is no negative ordinal; will have to be updated if we
+      // add non-stateless strategies (for which we've reserved negative ordinals; see
+      // Strategy.ordinal()).
+      strategyOrdinal = din.readByte();
+      numHashFunctions = UnsignedBytes.toInt(din.readByte());
+      dataLength = din.readInt();
+
+      Strategy strategy = BloomFilterStrategies.values()[strategyOrdinal];
+      long[] data = new long[dataLength];
+      for (int i = 0; i < data.length; i++) {
+        data[i] = din.readLong();
+      }
+      return new BloomFilter<T>(new BitArray(data), numHashFunctions, funnel, strategy);
+    } catch (RuntimeException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)